core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala

```diff
@@ -433,28 +436,25 @@ final class ShuffleBlockFetcherIterator(
       address: BlockManagerId,
       blockInfos: Seq[(BlockId, Long, Int)],
       collectedRemoteRequests: ArrayBuffer[FetchRequest]): Unit = {
-    val iterator = blockInfos.iterator
     var curRequestSize = 0L
-    var curBlocks = Seq.empty[FetchBlockInfo]
+    val curBlocks = new ArrayBuffer[FetchBlockInfo]()
 
-    while (iterator.hasNext) {
-      val (blockId, size, mapIndex) = iterator.next()
+    for ((blockId, size, mapIndex) <- blockInfos) {
       assertPositiveBlockSize(blockId, size)
-      curBlocks = curBlocks ++ Seq(FetchBlockInfo(blockId, size, mapIndex))
+      curBlocks += FetchBlockInfo(blockId, size, mapIndex)
       curRequestSize += size
       // For batch fetch, the actual block in flight should count for merged block.
       val mayExceedsMaxBlocks = !doBatchFetch && curBlocks.size >= maxBlocksInFlightPerAddress
       if (curRequestSize >= targetRemoteRequestSize || mayExceedsMaxBlocks) {
-        curBlocks = createFetchRequests(curBlocks, address, isLast = false,
-          collectedRemoteRequests)
+        val rest = createFetchRequests(curBlocks, address, isLast = false, collectedRemoteRequests)
+        curBlocks.clear()
+        curBlocks ++= rest
```

Comment:
Instead of clearing it, create new instance ?

Reply:
> If code evolves to use curBlocks in createFetchRequests or downstream (as Seq)

I not sure it's a good idea to use an ArrayBuffer as Seq implicitly.

Reply:
The PR is already doing this for `mergeContinuousShuffleBlockIdsIfNeeded` ?

Reply:
> The PR is already doing this for `mergeContinuousShuffleBlockIdsIfNeeded`? 
Not by this PR. `mergeContinuousShuffleBlockIdsIfNeeded` takes both `Seq` for handling local/host-local blocks and `ArrayBuffer` for the remotes. This PR modified it to ensure that it always returns a new copy for both `doBatchFetch` and `!doBatchFetch`. And before this PR,  `!doBatchFetch` branch might potentially result in surprises.

Reply:
> The PR is already doing this for `mergeContinuousShuffleBlockIdsIfNeeded`? 

Not by this PR. `mergeContinuousShuffleBlockIdsIfNeeded` takes both `Seq` for handling local/host-local blocks and `ArrayBuffer` for the remotes. This PR modified it to ensure that it always returns a new copy for both `doBatchFetch` and `!doBatchFetch`. And before this PR,  `!doBatchFetch` branch might potentially result in surprises.

Reply:
If you insist, I can revert the last commit https://github.com/apache/spark/pull/33063/commits/f8942209f2a00ef32940cd2805474c117fd89d5e. I do not have a strong opinion on it.